File: tests/example-frames/deformers/blend-shapes.cpp
// $dep ufbx mushnub_evolved.fbx
// $ clang++ ufbx.c main.cpp -lm -o example
// $ ./example mushnub_evolved.fbx Blink=0.5 Smile=0.7 Spikes=0.4 > result.obj
#include "ufbx.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <vector>

// -- EXAMPLE_SOURCE --

struct BlendShape {
    std::string name;
};

// Non-indexed triangulated skinned mesh
struct Mesh {
    std::vector<BlendShape> blends;
    std::vector<Vertex> vertices;
};

Mesh process_blend_mesh(ufbx_mesh *mesh, ufbx_blend_deformer *blend)
{
    Mesh result;

    // Triangulate the mesh, using `get_blend_vertex()` to fetch each index.
    size_t num_tri_indices = mesh->max_face_triangles * 3;
    std::vector<uint32_t> tri_indices(num_tri_indices);
    for (ufbx_face face : mesh->faces) {
        uint32_t num_tris = ufbx_triangulate_face(
            tri_indices.data(), tri_indices.size(), mesh, face);
        for (size_t i = 0; i < num_tris * 3; i++) {
            uint32_t index = tri_indices[i];
            result.vertices.push_back(get_blend_vertex(mesh, blend, index));
        }
    }
    assert(result.vertices.size() == mesh->num_triangles * 3);

    // Create bone descriptions
    size_t num_blends = blend->channels.count;
    if (num_blends > MAX_BLENDS) {
        num_blends = 0;
    }
    for (size_t i = 0; i < num_blends; i++) {
        result.blends.push_back(BlendShape{ blend->channels[i]->name.data });
    }

    return result;
}

struct Control {
    std::string name;
    float weight;
};

size_t g_vertex_offset = 0;

// Example implementation that dumps the skinned mesh as an .obj file
void dump_mesh(Mesh mesh, ufbx_node *node, const std::vector<Control> &controls)
{
    // Preload the control weights for the mesh blend shapes
    float weights[MAX_BLENDS] = { };
    size_t blend_ix = 0;
    for (const BlendShape &blend : mesh.blends) {
        for (const Control &control : controls) {
            if (control.name == blend.name) {
                weights[blend_ix] = control.weight;
                break;
            }
        }
        blend_ix++;
    }

    // Process each vertex
    for (const Vertex &vertex : mesh.vertices) {
        ufbx_vec3 p = vertex.position;
        for (size_t i = 0; i < MAX_BLENDS; i++) {
            ufbx_vec3 offset = vertex.blend_offsets[i];
            float weight = weights[i];
            p.x += offset.x * weight;
            p.y += offset.y * weight;
            p.z += offset.z * weight;
        }

        // The blend offsets are applied in local space, transform to global space
        p = ufbx_transform_position(&node->geometry_to_world, p);
        printf("v %.3f %.3f %.3f\n", p.x, p.y, p.z);
    }

    // Just a flat list of non-indexed triangle faces
    for (size_t vi = 0; vi < mesh.vertices.size(); vi += 3) {
        size_t i = g_vertex_offset + vi;
        printf("f %zu %zu %zu\n", i + 1, i + 2, i + 3);
    }
    g_vertex_offset += mesh.vertices.size();
}

int main(int argc, char **argv)
{
    ufbx_scene *scene = ufbx_load_file(argv[1], NULL, NULL);
    assert(scene);

    // Parse the blend shape controls, of form "Name=0.3"
    std::vector<Control> controls;
    for (int i = 2; i < argc; i++) {
        const char *arg = argv[i];
        const char *split = strchr(arg, '=');
        assert(split);
        controls.push_back(Control{
            std::string(arg, split), strtof(split + 1, NULL),
        });
    }

    for (ufbx_mesh *mesh : scene->meshes) {
        if (mesh->blend_deformers.count == 0) continue;
        ufbx_blend_deformer *blend = mesh->blend_deformers[0];

        Mesh result = process_blend_mesh(mesh, blend);

        for (ufbx_node *node : mesh->instances) {
            dump_mesh(result, node, controls);
        }
    }

    ufbx_free_scene(scene);
    return 0;
}

File: tests/example-frames/deformers/blend-shapes.rs
// $dep mushnub_evolved.fbx
// $ cargo build
// $ cargo run --quiet -- mushnub_evolved.fbx Blink=0.5 Smile=0.7 Spikes=0.4 > result.obj

// `Vertex::normal` is not used in this example
#![allow(unused)]

// -- EXAMPLE_SOURCE --

struct BlendShape {
    name: String,
}

// Non-indexed triangulated skinned mesh
struct Mesh {
    blends: Vec<BlendShape>,
    vertices: Vec<Vertex>,
} 

fn process_blend_mesh(mesh: &ufbx::Mesh, blend: &ufbx::BlendDeformer) -> Mesh {
    let mut vertices: Vec<Vertex> = Vec::new();

    // Triangulate the mesh, using `get_blend_vertex()` to fetch each index.
    let mut tri_indices = vec![0u32; mesh.max_face_triangles * 3];
    for face_index in 0..mesh.num_faces {
        let face = mesh.faces[face_index];

        // Triangulate the face into `tri_indices[]`.
        let num_tris = mesh.triangulate_face(&mut tri_indices, face);
        let num_tri_corners = (num_tris * 3) as usize;
        for &index in &tri_indices[..num_tri_corners] {
            vertices.push(get_blend_vertex(mesh, blend, index as usize));
        }
    }
    assert!(vertices.len() == mesh.num_triangles * 3);

    // Create bone descriptions
    let blends = blend.channels
        .iter()
        .map(|channel| BlendShape{
            name: channel.element.name.to_owned(),
        })
        .take(MAX_BLENDS)
        .collect();

    Mesh{ blends, vertices }
}

struct Control {
    name: String,
    weight: f32,
}

// Example implementation that dumps the skinned mesh as an .obj file
fn dump_mesh(mesh: &Mesh, node: &ufbx::Node, controls: &[Control], vertex_offset: &mut usize) {
    // Preload the control weights for the mesh blend shapes
    let mut weights = [0.0f32; MAX_BLENDS];
    for (i, blend) in mesh.blends.iter().enumerate() {
        for control in controls {
            if control.name == blend.name {
                weights[i] = control.weight;
            }
        }
    }

    // Process each vertex
    for vertex in &mesh.vertices {
        let mut p = vertex.position;
        for i in 0..MAX_BLENDS {
            let offset = vertex.blend_offsets[i];
            let weight = weights[i] as f64;
            p.x += offset.x * weight;
            p.y += offset.y * weight;
            p.z += offset.z * weight;
        }

        // The blend offsets are applied in local space, transform to global space
        p = ufbx::transform_position(&node.geometry_to_world, p);

        println!("v {:.3} {:.3} {:.3}", p.x, p.y, p.z);
    }

    // Just a flat list of non-indexed triangle faces
    for tri in 0 .. mesh.vertices.len() / 3 {
        let i = *vertex_offset + tri * 3;
        println!("f {} {} {}", i + 1, i + 2, i + 3);
    }
    *vertex_offset += mesh.vertices.len();
}

fn main() {
    let path = std::env::args().nth(1).expect("expected .fbx path");

    let scene = ufbx::load_file(&path, Default::default()).expect("failed to load scene");

    // Parse the blend shape controls, of form "Name=0.3"
    let controls: Vec<Control> = std::env::args()
        .skip(2)
        .map(|arg| {
            let (name, value) = arg.split_once('=')
                .expect("expected a Blend=(amount) argument");
            let name = name.to_owned();
            let weight: f32 = value.parse()
                .expect("expected blend value to be a float");
            Control{ name, weight }
        })
        .collect();

    let mut vertex_offset = 0;
    for mesh in &scene.meshes {
        if mesh.blend_deformers.len() == 0 {
            continue;
        }
        let blend = &mesh.blend_deformers[0];

        let result = process_blend_mesh(mesh, blend);

        for node in &mesh.element.instances {
            dump_mesh(&result, node, &controls, &mut vertex_offset);
        }
    }
}
